compiler/rustc_codegen_ssa/src/mir/mod.rs

```diff
@@ -171,6 +172,32 @@ pub fn codegen_mir<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     let fn_abi = cx.fn_abi_of_instance(instance, ty::List::empty());
     debug!("fn_abi: {:?}", fn_abi);
 
+    if cx.tcx().codegen_fn_attrs(instance.def_id()).flags.contains(CodegenFnAttrFlags::NAKED) {
+        let cached_llbbs = IndexVec::new();
+
+        let fx: FunctionCx<'_, '_, Bx> = FunctionCx {
+            instance,
+            mir,
+            llfn,
+            fn_abi,
+            cx,
+            personality_slot: None,
+            cached_llbbs,
+            unreachable_block: None,
+            terminate_block: None,
+            cleanup_kinds: None,
+            landing_pads: IndexVec::from_elem(None, &mir.basic_blocks),
+            funclets: IndexVec::from_fn_n(|_| None, mir.basic_blocks.len()),
+            locals: locals::Locals::empty(),
+            debug_context: None,
+            per_local_var_debug_info: None,
+            caller_location: None,
+        };
```

Comment:
Why do you need to create an entire FunctionCx for this? You can generate `String` containing assembly and then pass this to the codegen backend, right? That may allow reusing that function in cg_clif to replace it's existing naked function implementation, which doesn't use `FunctionCx`.

Reply:
I added it to support `asm_const` https://github.com/rust-lang/rust/issues/93332. specifically some constants were not evaluated by the time they were needed to insert them into the template. I'm open to other solutions of course, but some amount of evaluation is apparently needed at this point.

Reply:
You don't need `FunctionCx` to do evaluation of constants. The full implementation of `eval_mir_constant` is https://github.com/rust-lang/rust/blob/2e6fc425411280c658d852813de01c5157255088/compiler/rustc_codegen_ssa/src/mir/constant.rs#L27-L29 Where `self.monomorphize()` is: https://github.com/rust-lang/rust/blob/2e6fc425411280c658d852813de01c5157255088/compiler/rustc_codegen_ssa/src/mir/mod.rs#L124-L128 You can inline both functions to get rid of the `FunctionCx` dependency.